libsolidity/analysis/ViewPureChecker.cpp

```diff
@@ -385,6 +385,18 @@ void ViewPureChecker::endVisit(MemberAccess const& _memberAccess)
 		auto const& type = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type);
 		MagicMember magicMember(type.kind(), member);
 
+		if (
+			m_currentFunction &&
+			m_currentFunction->isReceive() &&
+			type.kind() == MagicType::Kind::Message &&
+			member == "data"
+		)
+			m_errorReporter.typeError(
+				7139_error,
+				_memberAccess.location(),
+				R"("msg.data" cannot be used inside of "receive" function.)"
+			);
```

Comment:
Should it be type error? I am not entirely sure about that

Reply:
I'd say that's fine... Actually, I might have put this in the ``TypeChecker`` - it's not really a question of state mutability, is it?

Reply:
I guess that would be more involved, since the current function is not that readily available there? ``StaticAnalyzer`` would also be an option...

Reply:
Yep, StaticAnalyzer is certainly a better place.

Reply:
`StaticAnalyzer` has similar checks, I added it there. If you think it will fit better to `TypeChecker`, I can move it over there. It is not such a big deal to add current function over there.